Update device trust docs

flyinghermit opened the pull request:
Updates device trust docs:
-
Most of the work in this PR includes shuffling existing content to various new pages so that contents align with day 1 and day 2 usage of device trust. 
- **Base device trust page**: Now explains basic concepts in detail. Contents are  borrowed from outgoing "Set up Device Trust" page.
- **Getting started guide** (previously "Set up Device Trust" page): updated to focus on day 1 usage of device trust.
- **Device Management**: New page. Most of the contents borrowed from outgoing "Set up Device Trust" page. This page is focused on device management i.e. register device, enroll device, remove device, advanced config
- **Enforcing Device Trust**: New page focused on device trust enforcement, including locking device. Most of the contents borrowed from outgoing "Set up Device Trust" page. 
- **Auto enrollment**: Page is deleted and the content is added to **Device Management**. After reading my current edits (and also partly from the feedback that it was not easy to find the auto enrollment doc), I felt like it would be better for readers to know about auto enrollment when they are already on the Device Management page. But lmk if this is a wrong idea and will put it back. 
- **Jamf Integration**: Updated to add reference to hosted Jamf plugin. 

I also recommend to pull these changes and view them in local docs server to review if the links and docs flow is consumable when surfing Device Trust docs.

flyinghermit committed:
add dedicated page for device management and enforcement
- remove auto enrollment page
- update config.json

flyinghermit committed:
update guide page to getting started, update copy to focus on day 1 usage

flyinghermit committed:
update base device-trust doc with concepts.
- content is borrowed and expanded from existing doc

flyinghermit committed:
update copy and links

flyinghermit committed:
fix links, run linter write

flyinghermit committed:
add reference to hosted Jamf plugin

zmb3 commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,381 @@
+</Tabs>
+
+<Admonition type="warning" title="Device Role">
+For clusters created using Teleport v12 or newer, the builtin role editor has the necessary permissions to manage devices.
```
```suggestion
For clusters created using Teleport v12 or newer, the preset `editor` role has the necessary permissions to manage devices.
```

zmb3 commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,381 @@
+
+<Admonition type="warning" title="Device Role">
+For clusters created using Teleport v12 or newer, the builtin role editor has the necessary permissions to manage devices.
+If you Teleport cluster is older than v12, you will need to create and grant `device` role to manage device inventory.[link to below]
```
Don't forget to update this `[link to below]`

sshahcodes replied:
Thanks for spotting that. I swear I had updated that once but it must have resurfaced during copying content from docs preview to this PR branch 😭

zmb3 commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,381 @@
+Teleport supports auto-enrollment, which if enabled, automatically enrolls user
+device during next Teleport login.
+
+Auto-enrollment allows `tsh` to automatically enroll devices already registered
```
Is it Auto-enrollment or auto enrollment?

sshahcodes replied:
Currently, its "auto-enrollment". Though in my heading, I have "auto enrollment". I will update it to auto-enrollment for consistency.

zmb3 commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,381 @@
+integration, like the [Jamf Pro integration](./jamf-integration.mdx).
+
+<Admonition type="note" title="Device Registration">
+A previously registered device is necessary for
```
Didn't we already cover this in the previous paragraph?

sshahcodes replied:
Yeah this note does not seem to add any new, removing it.

zmb3 commented on docs/pages/access-controls/device-trust/guide.mdx:
```diff
@@ -1,462 +1,171 @@
-private key that is also protected by the TPM. The signed attestation ensures
-that the Teleport Auth Server knows both the state of the device and that the
-request has come from the device.
+  - User devices: macOS and Windows OS.
```
```suggestion
  - User devices: macOS and Windows.
```

zmb3 commented on docs/pages/access-controls/device-trust/guide.mdx:
````diff
@@ -1,462 +1,171 @@
 ```diff
-kind: cluster_auth_preference
-version: v2
+$ vi access.yaml
````
```suggestion
```

zmb3 commented on docs/pages/access-controls/device-trust/jamf-integration.mdx:
```diff
@@ -53,6 +51,17 @@ Create a readonly Jamf user for inventory sync.
 ## Step 2/4. Configure Jamf service
 
+<Admonition type="tip" title="Hosted Jamf plugin">
+  Teleport cloud users can quickly get started with Jamf service by configuring
```
```suggestion
  Teleport Cloud users can quickly get started with Jamf service by configuring
```

codingllama commented:
I'll take a closer look soon, but I just wanted to jump in and mention the recently-opened PRs #30865 and #30878, both of which might make things a bit simpler here.

sshahcodes commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,200 @@
+  Kubernetes connections.
+
+### Prerequisites
+(!docs/pages/includes/commercial-prereqs-tabs.mdx!)
```
added signed `tsh` prereqs here.

flyinghermit committed:
resolve review comments:
- multiple grammar update
- auto enrollment -> auto-enrollment
- add signed prereqs in enforcing ddevice trust page

sshahcodes commented:
> PRs https://github.com/gravitational/teleport/pull/30865 and https://github.com/gravitational/teleport/pull/30878, both of which might make things a bit simpler here.

Nice! That's a great addition 🚀

**Note to reviewer:** I will be updating this PR to address changes added in https://github.com/gravitational/teleport/pull/30865 and https://github.com/gravitational/teleport/pull/30878

flyinghermit committed:
accomodate new preset roles and default auto-enrollment:
- udpated getting started guide to use preset require-trusted-device role
- minor grammar and code formating

sshahcodes commented:
Friendly ping for review please!

flyinghermit committed:
fix typo udpate -> update

ptgott commented:
Hi @sshahcodes, to help me review this, would you mind explaining the problems this PR is solving? The summary in the description is helpful for understanding the changes at a high level, but I'm still not really sure why these changes are being made. Thanks!

sshahcodes commented:
@ptgott sure, the main reason for the changes in this PR is to adjust existing contents focused on day 1 (for users approaching device trust for the first time) and day 2 (for detailed explanation of available configurations, advance use cases, learn about edge case, best practice etc) docs. 

Currently, it's a bit mix of reference manual vs getting started all added to a single page without focus on reader persona. It was written when the feature was released first. Lots of defaults have been changed since then, including the new ones that's addressed in https://github.com/gravitational/teleport/pull/30865 and https://github.com/gravitational/teleport/pull/30878. Some of the prerequisite aren't even required anymore. The content was structured for general use case but most of it would not be necessary for a admin to quickly try Device Trust, for example, knowing advance use cases and setting up roles and permissions for secure operations is not necessary at all for day 1. 

This was highlighted in this ticket [improving Device Trust UX](https://github.com/gravitational/teleport/issues/30386 ). Some of the issues mentioned there would have been addressed if the docs pointed to right direction, hence the changes in this PR. Most of the content is reused from "Set up device trust" page and added to the following pages. 
**Day 1:** Base device trust page (explain concept), Getting started guide (new content)
**Day 2:** Device Management, Enforcing Device Trust, Jamf Integration

lsgunn-teleport reviewed (approved):
Overall, this looks good. I might suggest some languages tweaks in a follow-on PR but I think the re-org makes sense. My only question given the current structure is why you wouldn't just but device trust concepts in with the other Device trust topics?
Device trusts
- concepts
- get-started (how-to enroll a device - simple demo/eval/test/day 1 task)
- manage devices (how-to topics)
- enforce-trust (how-to topics)

flyinghermit committed:
remove auto-enrollment default enabled admonition

sshahcodes commented:
@lsgunn-teleport Thank you for the review. 

> I might suggest some languages tweaks in a follow-on PR

Sure, you can also create an issue ticket and tag me there, I can collaborate to improve Device Trust docs. 

> ... device trust concepts in with the other Device trust topics?

Nothing concrete in particular but given docs re-org, I wanted to keep the new content minimum as there would be too many things to review. I also feel how-to topics needs proper time to find right balance on content, steps and context and so adding reference style docs, as it is in current PR was helpful in quickly grouping relevant commands together while the concepts are explained in single page. But yes, definitely something to revisit in future to add more how-to guides.

sshahcodes commented:
Friendly ping for review please @zmb3 and @codingllama

codingllama commented:
Apologies for the delay, taking a look now.

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,200 @@
+kind: role
+version: v6
+metadata:
+  name: example-role
```
Let's use/reference `require-trusted-device` here. I don't think "example-role" is a good role name for the docs.

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+<Admonition type="tip" title="v13.3.6+">
+  Teleport version 13.3.6 and above supports preset `require-trusted-device` role to enforce Device Trust.
+  You have an option to either update existing role as referenced below, or use the preset role.
+  Ensure to update rules defined in preset `require-trusted-device` role as required.
```
```suggestion
  Make sure you update the "allow" rules in the role according to your requirements.
```

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
````diff
@@ -0,0 +1,208 @@
+```
+
+```code
+$ tctl create -f admin.yaml
````
This seems out of place, what is "admin.yaml" ?

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
````diff
@@ -0,0 +1,208 @@
+```code
+$ tctl create -f admin.yaml
+```
+
````
nit:

```suggestion
```

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+`required` will enforce a trusted device for all SSH, Database and Kubernetes
+accesses.
+</Admonition>
+
```
nit:

```suggestion
```

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+</Admonition>
+
+
+<Admonition type="warning" title="Web UI">
```
Double-admonition.

sshahcodes replied:
The web ui is important one but I think the admonition block above this "Enterprise clusters run in `optional` mode by default. Changing the mode to `required` will enforce a trusted device for all SSH, Database and Kubernetes
accesses." can be added as a normal sentence.

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+leaf clusters.
+</Admonition>
+
+<Admonition type="tip" title="Per-role vs. Cluster-wide">
```
Double-admonition.

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+</Admonition>
+
+<Admonition type="tip" title="Per-role vs. Cluster-wide">
+Per-role configuration let's you progressively block unauthorized device access,
```
```suggestion
Per-role configuration lets you progressively block unauthorized device access,
```

codingllama commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+and is useful for getting started with Teleport Device Trust.
+
+Once all the devices used are enrolled in Teleport, cluster-wide
+configuration will ensure every resources can only ever be accessed with a
```
```suggestion
enforcement will ensure every resources can only ever be accessed with a
```

zmb3 commented:
Lot's of great feedback here already, so I'll wait to review until this batch is addressed.

sshahcodes commented on docs/pages/access-controls/device-trust.mdx:
```diff
@@ -1,31 +1,97 @@
 ---
 title: Device Trust (Preview)
-description: Use and enforce trusted devices with Teleport
+description: Teleport Device Trust Concepts
```
We use **"Teleport"** when referring to product names. E.g. Teleport Server Access, Teleport Database Access. Few example below: https://github.com/gravitational/teleport/blob/166146287276a3395d1d8f39a288a8955ab0c575/docs/pages/desktop-access/introduction.mdx#L3

https://github.com/gravitational/teleport/blob/166146287276a3395d1d8f39a288a8955ab0c575/docs/pages/server-access/introduction.mdx#L2

https://github.com/gravitational/teleport/blob/166146287276a3395d1d8f39a288a8955ab0c575/docs/pages/database-access/getting-started.mdx#L3

sshahcodes commented on docs/pages/access-controls/device-trust.mdx:
```diff
@@ -1,31 +1,97 @@
+## Device Trust enforcement
+
+Enforcing Device Trust means configuring Teleport with device trust mode, i.e. applying
+`device_trust_mode: required` rule, which tells Teleport Auth Service to only allow access
```
Thanks for flagging that. I think the preferred term is "Teleport Auth Service" but I forgot to update "Teleport Auth Server" which I copied from existing doc. Updating all to  "Teleport Auth Service"

sshahcodes commented on docs/pages/access-controls/device-trust/guide.mdx:
````diff
@@ -1,462 +1,212 @@
+  - ```code
+    $ tsh status
+    > Profile URL:      https://teleport.example.com:443
+    Logged in as:       myuser
````
"myuser" is equally used in our existing docs and reads better over "alice". But I came to this name as the server access guides use "myuser" which I think has a relevant connection to this guide as we tell user to SSH. 

Though I didn't find a predefined variable for it so handcoded here, keeping as is.

sshahcodes commented on docs/pages/access-controls/device-trust/guide.mdx:
```diff
@@ -1,462 +1,212 @@
+    $ tsh status
+    > Profile URL:      https://teleport.example.com:443
+    Logged in as:       myuser
+    Cluster:            teleport.example.com
```
Funny I couldn't find it. Can you please share the link to the var definition if you find one.

sshahcodes replied:
So I have added default sets of variable in https://github.com/gravitational/teleport/pull/30821/commits/deaa15ec835baebb0b3c4e6d0c6f86f594d5da35

But since there are variations in the names we use in other docs, and not to blow the scope of this PR, I have also opened a ticket to discuss that https://github.com/gravitational/teleport/issues/31207

sshahcodes commented on docs/pages/access-controls/device-trust/guide.mdx:
```diff
@@ -1,462 +1,212 @@
-  <TabItem label="macOS">
-    The serial number of a macOS device can be determined in two ways.
+  <Details title="Update a role to enforce Device Trust" closed>
+    For simplicity, the example below updates the preset `access` role but you can update
```
> Updating access could have wide-ranging effects.

For pro (already using Teleport) users , yes and I think that group probably understand the implication and will go about updating/creating another role anyway. But users just starting with Teleport, I don't think it will have that serious impact and guiding to create another role just adds fricition while getting started. 

If you think the current copy does not put much emphasis on the risk, I had the following copy before:
"For simplicity, this guide works on the default preset `access` role. If other users in
your cluster are already assigned with this `access` role to access resources, and you do
not want to enforce device trust for them yet, create a separate `device-enforcment`
role and use that role for testing to avoid disruption."

Let me know what you think we should at least use the above copy (or similar variation). But I don't think its a good idea to tell users to create another role for day 1 usage.

codingllama replied:
I like your suggested copy, but feel free to do what you think is best.

sshahcodes replied:
I'll keep this as is and raise my hand if angry user shows up in the future complaining this guide blew off their team's access.

sshahcodes commented on docs/pages/access-controls/device-trust/guide.mdx:
```diff
@@ -1,462 +1,212 @@
-Once all the devices used are enrolled in Teleport, cluster-wide
-configuration will ensure every resources can only ever be accessed with a
-trusted device.
+<Admonition type="note" title="Windows">
```
Good point. I think it's okay to remove the Windows admonition here as it is not really required to be mentioned here.

sshahcodes commented on docs/pages/access-controls/device-trust/device-management.mdx:
````diff
@@ -0,0 +1,388 @@
+  ```
+</Admonition>
+
+<Admonition type="tip" title="Sync with Jamf Pro">
````
Hmm, I don't think we have much choice here as we want to mention both "easy alternative" before starting manual operation. Both needs to be mentioned in `tip` and does not makes sense introducing later in the docs. Introducing Jamf can be mentioned later but my gut tells it to introduce before we tell user to register device.

sshahcodes commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,388 @@
+a device, you first need to determine its serial number.
+
+<Tabs>
+  <TabItem label="macOS">
```
Yeah I totally missed to introduce `tsh device asset-tag` command. 

> `tctl devices add --current-device`

I am not quiet sure about its usage though. Can you please explain the context when would one use `tctl devices add --current-device` over `tsh device enroll --current-device`?

sshahcodes replied:
I know that difference but what I meant to ask was that given a user with required permissions and access to the device they want to register, why would they prefer to run `tctl devices add --current-device` over `tsh device enroll --current-device`.

codingllama replied:
The latter is always more practical, if that's what you are asking. `tctl` just doesn't have the capabilities to do what `tsh device enroll` does.

sshahcodes commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,388 @@
+successfully enrolled and authenticated. The device above is now a trusted device.
+
+<Admonition type="note" title="Windows">
+On Windows, the `tsh device enroll` command attempts to elevate to
```
Yeah i don't think it needs to be explained at all.

sshahcodes commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,388 @@
+Teleport supports auto-enrollment, which if enabled, automatically enrolls user
+device during next Teleport login.
+
+Auto-enrollment allows `tsh` to automatically enroll devices already registered
+in Teleport during the user's login. The registration may be
+[manual](#register-a-trusted-device) or performed using an
+integration, like the [Jamf Pro integration](./jamf-integration.mdx).
```
Reading now after updating paragraph before this (https://github.com/gravitational/teleport/pull/30821#discussion_r1309164628) I don't think this paragraph adds any new thing. 

Updating it to be:
`A device must be registered for auto-enrollment to work. Registration may be
[manual](#register-a-trusted-device) or performed using an
integration, like the [Jamf Pro integration](./jamf-integration.mdx).`

sshahcodes commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,388 @@
+
+The presence of the `teleport-device-*` extensions shows that the device was
+successfully enrolled and authenticated.
+
```
Hmm, I always thought (never tried beyond `editor` role) that auto enrollment works for any user, as long as:
1. Auto-enrollment is enabled
2. Device is registered.

Is this new addition or this was always the case?

codingllama replied:
It was always the case. We can make it _not_ the case, but I think the system gets a lot weaker as a consequence. Up to product, feel free to raise it.

sshahcodes replied:
That makes sense. Updated doc accordingly in https://github.com/gravitational/teleport/pull/30821/commits/bffb276385a8ce48f853846d7370ec9f6279bf9c

sshahcodes commented on docs/pages/access-controls/device-trust/device-management.mdx:
````diff
@@ -0,0 +1,388 @@
+role 'device-admin' has been created
+```
+
+Note that in addition to the usual CRUD verbs (create, read, list, update and
+delete), we have also included `create_enroll_token` and `enroll`. The
+`create_enroll_token` verb is necessary to execute the `tctl devices enroll`
+command; `enroll` is necessary to execute `tsh device enroll`. Both are
+demonstrated in the following section.
+
````
That needs to be removed :) 

Yes, this and lots of paragraph above are just moved from the original "set up device trust" doc. I have added very little new copy (except for the guide)

sshahcodes commented on docs/pages/access-controls/device-trust/device-management.mdx:
```diff
@@ -0,0 +1,388 @@
+demonstrated in the following section.
+
+
+## Configuring a TPM EKCert CA allow-list
```
Clean move.

sshahcodes commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+
+Once all the devices used are enrolled in Teleport, cluster-wide
+configuration will ensure every resources can only ever be accessed with a
+trusted device.
```
I am removing that entire comparison admonition block.

sshahcodes commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
```diff
@@ -0,0 +1,208 @@
+</Admonition>
+
+
+## Locking a device
```
I think it belongs in enforcement but tagging @zmb3 for third opinion.

sshahcodes commented on docs/pages/access-controls/device-trust/guide.mdx:
```diff
@@ -1,462 +1,212 @@
-A role that grants permissions for the `device` resource is necessary to manage
-the inventory. Save the yaml below as `device-admin.yaml` and create it in your
-cluster:
+In this guide, you will update an existing user profile to assign the preset `require-trusted-device`
```
The paragraph above this explain the base generic steps:
```
Device Trust requires two of the following steps to have been configured:

- Device enforcement mode configured via either a role or a cluster-wide config.
- Trusted device registered and enrolled with Teleport.
```
so IMO this one feels natural to read as a followup sentence. 

```
Device Trust requires two of the following steps to have been configured:

- Device enforcement mode configured via either a role or a cluster-wide config.
- Trusted device registered and enrolled with Teleport.

In this guide, you will update an existing user profile to assign the preset `require-trusted-device`
role and then enroll a trusted device into Teleport to access a resource (a test linux server)
protected with Teleport.
```

sshahcodes commented on docs/pages/access-controls/device-trust.mdx:
```diff
@@ -1,31 +1,97 @@
+- Trusted device registered and enrolled with Teleport.
+- Device enforcement mode configured via either a role or a cluster-wide config.
+
+Categorically, we define these two requirements as Trusted Device management
+and Device Trust enforcement.
```
I hear you on this one. Between "Trusted Device management/enforcement" and "Device Trust management/enforcement", I prefer the first one as it is close to how device trust works(enrolling trusted devices, enforcing trusted device access). 

Though the Device Trust Enforcement also aligns with the configuration semantics as the rule to enforce device is `device_trust_mode: required` and not `trusted_device: required`. That said, I am open to suggestions here. 

Also tagging @zmb3 for more suggestions.

flyinghermit committed:
address review suggestions:
- adds resolved comments

flyinghermit committed:
add clusterDefaults variable for generic code reference

flyinghermit committed:
add troubleshooting to guide and enforcment page

sshahcodes commented on docs/pages/access-controls/device-trust/guide.mdx:
````diff
@@ -1,462 +1,212 @@
     ```code
-    $ ioreg -c IOPlatformExpertDevice -d 2 | grep -i IOPlatformSerialNumber | awk -F'"' '{print $4}'
-    (=devicetrust.asset_tag=)
+    $ tctl get roles/access > access.yaml
````
agree `tctl edit` is much better reference here.

flyinghermit committed:
reference tctl edit role/access and add tsh relogin

flyinghermit committed:
fix typo

sshahcodes commented:
Friendly ping @codingllama @zmb3. 
I have addressed most of the previous suggestions but few are waiting for discussion.

codingllama commented on docs/pages/access-controls/device-trust/device-management.mdx:
````diff
@@ -36,18 +36,41 @@ are automatically updated in Teleport's device inventory.
+(=devicetrust.asset_tag=)
+```
+
+<Details title="Manually retrieving device serial" closed>
````
I like how simpler this got after `tsh device asset-tag`. Nice.

codingllama reviewed (approved):
Thanks for tackling this one.

zmb3 commented on docs/pages/access-controls/device-trust/enforcing-device-trust.mdx:
````diff
@@ -0,0 +1,191 @@
+Lock a device:
+
+```code
+$ tctl lock --device (=devicetrust.device_id=) --ttl=12h
````
```suggestion
$ tctl lock --device=(=devicetrust.device_id=) --ttl=12h
```

flyinghermit committed:
mention required role for auto-enrollment and minimum requirements

flyinghermit committed:
add  in  command

flyinghermit committed:
fix typo
